fix: do not double handle errors in component testing

lmiller1990 opened the pull request:
<!-- Thanks for contributing! PLEASE...
- Read our contributing guidelines: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md 
- Read our Code Review Checklist on coding standards and what needs to be done before a PR can be merged: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#Code-Review-Checklist
- Mark this PR as "Draft" if it is not ready for review.
- Make sure you set the correct base branch based on what packages you're changing: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md#branches
-->

- Closes https://github.com/cypress-io/cypress/issues/23920

### User facing changelog
<!-- 
Explain the change(s) for every user to read in our changelog. Examples: https://on.cypress.io/changelog
If the change is not user-facing, write "n/a".
-->

Fix a bug where unhandled exceptions and errors would be logged twice in Component Testing.

### Additional details
<!-- Examples:
- Why was this change necessary?
- What is affected by this change?
- Any implementation details to explain?
-->

Unlike End To End Testing which has two iframes:
- Spec Frame, for the spec.
- AUT Frame, for the user's application.

Component Testing only has one iframe:
- AUT Frame

AUT Frame is _also_ the Spec Frame, since we serve the specs from a dev server - they are bundled as a single file.
We don't want to bind two error handlers, or we end up logging errors twice. For this reason, we conditionally add these event listeners here - Component Testing errors are captured and logged in `contentWindowListeners#bindToListeners` in `cypress/cy.ts`.

### Steps to test
<!--
For non-trivial behavior changes, list the steps that a reviewer should follow to validate the new behavior.
This is not meant to be the only testing performed by a reviewer, just the "happy path" that leads to the new behavior.
-->

See steps to reproduce in https://github.com/cypress-io/cypress/issues/23920. Make a component that errors when you clic a button, you can see that before we'd log the same error multiple times - now we do not.

Also, you could copy+paste my regression test (added in driver) into develop and watch it fail.

### How has the user experience changed?
<!-- Provide before and after examples of the change.
Screenshots or GIFs are preferred. -->

### PR Tasks
<!-- 
These tasks must be completed before a PR is merged.
If a task does not apply, write [na] instead of checking the box.
DO NOT DELETE the PR checklist.
-->

- [x] Have tests been added/updated?
- [ ] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? <!-- Link to PR here -->
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

lmiller1990 committed:
fix: do not double catch errors and unhanded exceptions in component testing

lmiller1990 committed:
add test fixtures

lmiller1990 committed:
revert whitespace

lmiller1990 committed:
snaps

lmiller1990 commented on system-tests/__snapshots__/vite_dev_server_fresh_spec.ts.js:
```diff
@@ -391,11 +395,15 @@ We dynamically generated a new test to display this failure.
 
   1) is missing React:
-     ReferenceError: The following error originated from your test code, not from Cypress.
+     ReferenceError: The following error originated from your application code, not from Cypress.
```
Weird one - I think application is more accurate here. In this example the user wrote a JSX file but didn't do `import React from 'react'` - I'd say that's their application code.

For CT, application/spec is really a blurred line - it's one bundle and one iframe, so it's kind of the same thing.

ZachJW34 replied:
Yeah a bit weird, I added a `setTimeout(() => throw new Error('wtf))` inside an `it` and the error is the same (type=application). Not sure of a way to differentiate between app and test code when it comes to CT unless you looked at the error stack.

I would be in favor of it always saying `test` code for CT

lmiller1990 replied:
We could do this, we'd need to move this conditional error listener. It's hard to say if it's your component code or your test code, since there is only one frame in CT - in E2E, it's much more clear, since there's the AUT and Spec Frames.

One option would be to make more generic for CT - something like "The following error originated from your code"?

lmiller1990 replied:
Actually, I thought about it more - this is probably worth spending a bit of time on, it's quite a foundational part of Cypress. I am not sure if there is a good way to know if it's spec or app error, though. I'll take a look.

lmiller1990 replied:
Ok, something like

1. Get stack trace
2. Get origin of error
3. If spec file -> spec error
4. else -> App error

This means we need to see if the file in the stack traces matches the specPattern glob. This is tricky.

ZachJW34 replied:
What happens if the error occurs on the support file? Like you mentioned earlier, I'm not sure it makes sense to make the distinction between app and test code for CT. You could have a cypress utils file that would be considered "App" code, or a Component Harness only used for CT (but since the code doesn't come from a `cy.{js,ts}` file it is considered "App" code).

The distinction between test code and app code for e2e is perfectly defined, not so much with CT so I'm not sure it's worth spending too much time on this if the distinction can't be made.

lmiller1990 replied:
Yeah I agree, maybe I'll spend 1 more hour or so. Basic heuristic could be

- spec file OR support file: test code
- anything else: application code

If this is too much work, I'll just pick one and go with it - either way, the stack trace shows, which is the main thing.

Either way, while working on this, I found a few more edge case, going to add some more tests.

lmiller1990 replied:
@ZachJW34 I thought about this more. I compiled a list of every error and how the UI looks after this change: https://www.notion.so/cypressdx/Errors-adf3361e7c174c388404142148a865a5

WDYT? "Application Error" seems more accurate that "Spec Error" for the majority of cases.

ZachJW34 commented on packages/driver/cypress.config.ts:
```diff
@@ -20,4 +21,14 @@ export default defineConfig({
+  component: {
+    specPattern: 'cypress/component/**/*.cy.js',
+    supportFile: false,
+    devServer: (devServerOptions) => {
```
Any reason to use the function over the object API?

lmiller1990 replied:
This package (driver) doesn't have any component tests (there aren't any components to be tested). 

If you use the object API, you need to specify `framework` - and we don't have one here.

Unless it's possible to just use `bundler` without `framework`? I don't think we support this, do we?

ZachJW34 replied:
The types might not be happy but you should be able to do:
```js
devServer: {
  bundler: 'webpack',
  framework: undefined
} as any
```

mike-plummer reviewed (approved):
LGTM 🚀

lmiller1990 committed:
better types

lmiller1990 committed:
Merge remote-tracking branch 'origin/develop' into lmiller/issue-23920

lmiller1990 commented:
Moving this back to in progress - there's a few more things we need to do, eg - verify the right error message ("error is from your test code" vs "error is from your application code", etc).

lmiller1990 committed:
improve types

lmiller1990 committed:
do not double capture errors

lmiller1990 committed:
add test

lmiller1990 committed:
adding tests

lmiller1990 committed:
updates

lmiller1990 committed:
revert

lmiller1990 committed:
update

lmiller1990 committed:
types

lmiller1990 commented on packages/driver/src/cypress/error_utils.ts:
```diff
@@ -384,16 +385,28 @@ const errByPath = (msgPath, args?) => {
-const createUncaughtException = ({ frameType, handlerType, state, err }) => {
+type FrameType = 'spec' | 'app'
+
+interface UncaughtException {
```
Slowly typing things!

lmiller1990 committed:
give up on types, it is impossible

lmiller1990 committed:
snaps

lmiller1990 committed:
fix

lmiller1990 committed:
Merge branch 'develop' into lmiller/issue-23920

lmiller1990 commented:
@mike-plummer @ZachJW34 back in review. I dismissed your reviews - please take another look (code is mostly the same, just added some tests and cleaned some things up).

I added more tests to ensure we don't double log errors, building on top of @mike-plummer stack trace tests. 

```js
verify('sync error', {
      fileName: 'Errors.vue',
      line: 24,
      column: 16,
      uncaught: true,
      uncaughtMessage: 'sync error',
      message: [
        'The following error originated from your application code',
        'sync error',
      ],
      codeFrameText: 'Errors.vue',
    }).then(() => {
        verifyErrorOnlyCapturedOnce('Error: sync error') // <==== this is new 
    })
})
```

Created a doc showing all the different cases: https://www.notion.so/cypressdx/Errors-adf3361e7c174c388404142148a865a5

There is also one outstanding bug: https://github.com/cypress-io/cypress/pull/23957/files#diff-0848e3c3d5abbc06e692c0d0ace10635a59161668729e1915c4042d227a65a27R94

ReactDOM seems to double-throw errors for some reason - only React component that error still double-register. I think we need to look into ReactDOM - it seems to do something different other libraries. I created a follow up issue: https://github.com/cypress-io/cypress/issues/24011

ZachJW34 reviewed (approved):
Thanks for putting together those error conditions @lmiller1990, and the additional tests are great! I'm good with the behavior

mike-plummer reviewed (approved):
LGTM except for one small cleanup task - are we able to remove this special case now that the duplicates aren't being logged? https://github.com/cypress-io/cypress/pull/23957/files#diff-29381ef9f5aca1b836d8ff0879ad99a5659a320230b2d74ec5665227de062b22L156 (`verify-failures.ts`, line 155-156)

I will approve assuming you respond/address the above comment

lmiller1990 committed:
Merge remote-tracking branch 'origin/develop' into lmiller/issue-23920

lmiller1990 commented:
@mike-plummer done, will merge once CI is green. ✅

lmiller1990 committed:
remove link
